### DISCUSSION TITLE
perspective transformation to map a quadrilateral to rectangle using image magick

### DISCUSSION THREAD

**Discussion by thisisnottaken** [ImageMagick/ImageMagick#2760] [Digital Image Processing]
hi,
Consider this case - I have a quadrilateral ABCD and I want to map it to rectangle EFGH .
See https://ibb.co/FwmHHJj
Mathematical details to do this can be found at http://web.archive.org/web/20100612191527/http://alumni.media.mit.edu/~cwren/interpolator/
New coordinates for each pixel in rectangular image is given by
x_new=(ax+by+c)/(gx+hy+1) , y_new=(dx+ey+f)/(gx+hy+1)
Values of a,b,c,d,e,f,g and h can be obtained by solving 8 equations using cooordinates of corners of quadrilateral in original and modified image.
Even after calculating a , b , c , d , e , f , g , h - I find that x_new and y_new are not integers so mapping a pixel from original image to modified image is hard to implement in digital image.
Is there any support for this transformation in image magick ?

Thanks.

**Comment by GeeMack** [comment]
ImageMagick can do a basic perspective distortion with a IMv6 command like this. (For IMv7 use "magick" instead of "convert".)

    convert inimage.png -distort perspective "Ax,Ay Ex,Ey  Bx,By Fx,Fy   Cx,Cy Gx,Gy   Dx,Dy Hx,Hy" outimage.png

Use your values for the x and y locations of all your points. So for your example image that would move the x,y coordinates of point A to the x,y coordinates you choose for point E. And so on for each of the other three points.

![perspective-transformation](https://user-images.githubusercontent.com/65475279/97085510-7afb9680-15e3-11eb-9bfb-c783f80c3fc9.png)
That is about the simplest example. There are ways to tweak a command like that to suit special requirements, etc. Find out more about ImageMagick's perspective distortion here...

[https://legacy.imagemagick.org/Usage/distorts/#perspective](url)

Edited to add:

Here's an actual example command with IMv6 run on Windows that draws a polygon followed by a command to apply the perspective distortion...

    convert -size 640x480 -fill none -stroke black -strokewidth 4 xc:white -draw "polygon 80,50 580,70 620,420 40,410" polygon.png

![polygon](https://user-images.githubusercontent.com/65475279/97087253-67096200-15ee-11eb-9cae-97cfd1112e80.png)

And to apply the distortion...

    convert polygon.png -distort perspective "80,50 10,10   580,70 630,10   620,420 630,470   40,410 10,470" rectangle.png

![rectangle](https://user-images.githubusercontent.com/65475279/97087266-886a4e00-15ee-11eb-8486-ff3c1f1da70d.png)

**Reply by thisisnottaken** [reply]
thanks

**Reply by thisisnottaken** [reply]
Can you provide a way to obtain coordinates of points of quadrilateral .
I am using image magick in windows.

Version: ImageMagick 7.0.11-4 Q16 x86 2021-03-20 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829910
Features: Cipher DPC Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms
 lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib

I tried to use imdisplay to view image but am unable to get coordinates of pixel on which mouse is hovering .
Currently I use mspaint to get coordinates which has worked so far.

Thanks.

**Reply by snibgo** [reply]
> Currently I use mspaint to get coordinates which has worked so far.

Then you can continue using that. Gimp can also do that task.

**Reply by thisisnottaken** [reply]
> ImageMagick can do a basic perspective distortion with a IMv6 command like this. (For IMv7 use "magick" instead of "convert".)
> 
> ```
> convert inimage.png -distort perspective "Ax,Ay Ex,Ey  Bx,By Fx,Fy   Cx,Cy Gx,Gy   Dx,Dy Hx,Hy" outimage.png
> ```
> 
> Use your values for the x and y locations of all your points. So for your example image that would move the x,y coordinates of point A to the x,y coordinates you choose for point E. And so on for each of the other three points.
> 
> ![perspective-transformation](https://user-images.githubusercontent.com/65475279/97085510-7afb9680-15e3-11eb-9bfb-c783f80c3fc9.png)
> That is about the simplest example. There are ways to tweak a command like that to suit special requirements, etc. Find out more about ImageMagick's perspective distortion here...
> 
> [https://legacy.imagemagick.org/Usage/distorts/#perspective](url)
> 
> Edited to add:
> 
> Here's an actual example command with IMv6 run on Windows that draws a polygon followed by a command to apply the perspective distortion...
> 
> ```
> convert -size 640x480 -fill none -stroke black -strokewidth 4 xc:white -draw "polygon 80,50 580,70 620,420 40,410" polygon.png
> ```
> 
> ![polygon](https://user-images.githubusercontent.com/65475279/97087253-67096200-15ee-11eb-9cae-97cfd1112e80.png)
> 
> And to apply the distortion...
> 
> ```
> convert polygon.png -distort perspective "80,50 10,10   580,70 630,10   620,420 630,470   40,410 10,470" rectangle.png
> ```
> 
> ![rectangle](https://user-images.githubusercontent.com/65475279/97087266-886a4e00-15ee-11eb-8486-ff3c1f1da70d.png)

Hi ,
I realized that my results are not rectangular .
I am using
magick --version 
Version: ImageMagick 7.0.11-4 Q16 x86 2021-03-20 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Visual C++: 192829910
Features: Cipher DPC Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raw rsvg tiff webp xml zip zlib

I used command
magick convert arch-clip.jpg -distort perspective "13,443 0,427   1268,389 1264,427   1251,16 1264,0   4,38 0,0" arch-rect.jpg
![arch-rect](https://user-images.githubusercontent.com/19471922/113639200-49105500-9696-11eb-820c-849225f01c8c.jpg)
![arch-clip](https://user-images.githubusercontent.com/19471922/113639207-4c0b4580-9696-11eb-9959-cd5dd9f84b0b.jpg)

I can't figure out where am I going wrong ?

**Reply by thisisnottaken** [reply]
In above post - first image is output and second image is input.

Note that there is still a little red strip at bottom of output and little space near right edge of image.
I was hoping that black boundary will be almost clipped from bottom and right like the top and left edge.

**Reply by fmw42** [reply]
First note that in IM use magick, not magick convert.  

Your issue is that -distort tries to make the output the same size as the input. You need to use the viewport to define how big to make the output.

`magick  arch-clip.jpg -set option:distort:viewport 1264x427+0+0 -distort perspective "13,443 0,427 1268,389 1264,427 1251,16 1264,0 4,38 0,0" arch-rect.jpg`

![arch-rect](https://user-images.githubusercontent.com/13458218/113640203-c18ffb00-962f-11eb-9c9a-958364da9854.jpg)

See https://legacy.imagemagick.org/Usage/distorts/#distort_viewport

**Reply by thisisnottaken** [reply]
That viewport option worked well .

Can it be said that previously I was transforming a part of image and coordinates 0,0 and 0,something caused left hand side and top of image to be cropped but the right hand and bottom part of image that surrounds the quadrilateral remain unchanged .

**Reply by snibgo** [reply]
As always, different approaches are possible. This is how I would do it.

The input image is 1280x450 pixels. With "-distort", the output will be the same size. So the bottom-left of the output will be at (0,449). What input coordinate do I want at the output bottom-left? (12,440) seems reasonable.

Similarly, I want (1,36) to move to top-left (0,0), and (1248,12) to move to top-right (1279,0), and (1266,384) to move to bottom-right (1279,449). Windows BAT syntax:
```
magick ^
  113639207-4c0b4580-9696-11eb-9959-cd5dd9f84b0b.jpg ^
  -distort perspective "12,440,0,449 1,36,0,0 1248,12,1279,0 1266,384,1279,449 " ^
  x.jpg
```
![x](https://user-images.githubusercontent.com/33812368/113703446-bf32ad00-96d2-11eb-9fb4-e99d778e2380.jpg)

**Reply by thisisnottaken** [reply]
I get that . You have chosen output coordinates such that quadrilateral gets transformed into rectangle covering the whole image.
BTW you have taken input coordinates as 12,440 , 1,36 , 1248,12 , and 1266,384 . The order of these points is clockwise . GeeMack had used anti clockwise direction of points in example . Are both allowed ?
Secondly my input coordinates were 13,443 , 4,38 , 1251,16 and 1268,389 . These differ from yours by few pixels. I guess you picked coordinates directly from image I posted instead of sourcing from my command .

**Reply by fmw42** [reply]
You can use any arrangement so long as the two sets (input and output) correspond. They do not need to be arranged either cw or ccw.

**Reply by snibgo** [reply]
Yes, I could have listed the points in any order.

Yes, I picked the input coords from the image.

**Comment by snibgo** [comment]
As GeeMack says.

> Even after calculating a , b , c , d , e , f , g , h - I find that x_new and y_new are not integers so mapping a pixel from original image to modified image is hard to implement in digital image.

IM does the mathematics so it can visit all the pixels in the output image and, from each coordinate, calculate the corresponding coordinate of the input image. Input coordinates will generally not be integers. To get the input colour, IM could simply round to the nearest integer. Or it could use an interpolation. The user can specify the method; see https://imagemagick.org/script/command-line-options.php#interpolate

**Reply by thisisnottaken** [reply]
thanks

**Comment by fmw42** [comment]
I use ImageJ and its pointpicker plugin. It will save the points to a file.
